webrtc/README.md

```diff
@@ -0,0 +1,416 @@
+  connections on? Does the browser reuse the UDP port across many WebRTC
+  connections? If that is the case one could connect to any public node, with
+  the remote telling the local node what port it is perceived on.
+
+  No, a browser uses a new UDP port for each `RTCPeerConnection`.
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+  Yes. For the lifetime of the page, one can generate a certificate once and
+  reuse it across connections. See also
+  https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection#using_certificates
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
+  based on it?
+
+- Instead of using trickle ICE, we could as well wait for the candidate
+  gathering. See
+  https://github.com/pion/webrtc/blob/c1467e4871c78ee3f463b50d858d13dc6f2874a4/examples/insertable-streams/main.go#L141-L142
+  as one example. In the browser, one can wait for the
+  [`icegatheringstatechange`
+  event](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/icegatheringstatechange_event).
+
+## Multiplexing
+
+The WebRTC browser APIs do not support half-closing of streams nor resets of the
+sending part of streams.
+[`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close)
+flushes the remaining messages and closes the local write and read side. After
+calling `RTCDataChannel.close()` one can no longer read from the channel. This
+lack of functionality is problematic, given that libp2p protocols running on top
+of transport protocols, like WebRTC, expect to be able to half-close or reset a
+stream. See [Connection Establishment in
+libp2p](https://github.com/libp2p/specs/blob/master/connections/README.md#definitions).
+
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
```

Comment:
I thought about the state transitions that this implies and came up with the following:

![PXL_20220929_061704073](https://user-images.githubusercontent.com/5486389/192954212-ad3da7f8-a5f0-40ed-b915-8b587f9856c8.jpg)

Is this the intended behaviour?

My reasoning was:

- Reset implies that the remote will no longer send data. This is equivalent to us sending `STOP_SENDING` except that we will discard any buffered data that we received so far.
- If we already sent `FIN`, receiving a reset means the stream is completely closed now as neither party will send or receive.
- Receiving a reset for a stream that we already send `STOP_SENDING` on is a no-op.

@mxinden I do have a language-specific question on top. In Rust, the only interface we have for streams is `AsyncRead` + `AsyncWrite`. In functional terms, we have:

- read
- write
- flush
- close

AFAICT, all implementations treat `close` as "close the write side" which makes sense given that it is part of `AsyncWrite`. It does however mean that we don't have the capability to close the reading side so any code around `STOP_SENDING` will be completely unused for Rust <-> Rust streams. Other languages might be different here so we still need to handle it if it is part of the spec of course but I just wanted to point this out.

We might want to consider adding a `CloseableAsyncRead` trait that is implemented for all our substreams where we allow users to close the reading side.

Reply:
> Reset implies that the remote will no longer send data. This is equivalent to us sending STOP_SENDING except that we will discard any buffered data that we received so far.

It’s the other way round, RESET means that *we* won’t be sending any more data on the stream.

Reply:
> > Reset implies that the remote will no longer send data. This is equivalent to us sending STOP_SENDING except that we will discard any buffered data that we received so far.
> 
> It’s the other way round, RESET means that _we_ won’t be sending any more data on the stream.

Sorry for the confusing wording. This was meant to read as "**Receiving** RESET is equivalent to us sending STOP_SENDING".

Reply:
I realised the diagram is probably confusing too. It pictures the state transitions for **sending** the `FIN` and `STOP_SENDING` flag and **receiving** the `RESET` flag.

Reply:
> @mxinden I do have a language-specific question on top. In Rust, the only interface we have for streams is `AsyncRead` + `AsyncWrite`. In functional terms, we have:
> 
>     * read
>     * write
>     * flush
>     * close
> 
> 
> AFAICT, all implementations treat `close` as "close the write side" which makes sense given that it is part of `AsyncWrite`. It does however mean that we don't have the capability to close the reading side so any code around `STOP_SENDING` will be completely unused for Rust <-> Rust streams. Other languages might be different here so we still need to handle it if it is part of the spec of course but I just wanted to point this out.


Correct @thomaseizinger, we don't have the capability to close the reading side and thus the active code path (sending a `STOP_SENDING`) will be unused in rust-libp2p.

> We might want to consider adding a `CloseableAsyncRead` trait that is implemented for all our substreams where we allow users to close the reading side.

I think that is a good idea. Though I suggest we only work on this once there is a need for it.

Reply:
I redrew the diagram:

![JPEG_20221011_213349_1426476781261431344](https://user-images.githubusercontent.com/5486389/195068165-002b434d-4b5b-4b42-93cd-fbe6c85d15be.jpg)

These are the state transitions now implemented in `rust-libp2p`.